stan/math/rev/fun/gp_periodic_cov.hpp

```diff
@@ -300,95 +38,106 @@ class gp_periodic_cov_vari<T_x, double, T_l, T_p> : public vari {
  * @throw std::domain_error if sigma <= 0, l <= 0, p <= 0, or
  *   x is nan or infinite
  */
-template <typename T_x>
-inline typename std::enable_if<
-    std::is_same<typename scalar_type<T_x>::type, double>::value,
-    Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic>>::type
-gp_periodic_cov(const std::vector<T_x> &x, const var &sigma, const var &l,
-                const var &p) {
-  const char *fun = "gp_periodic_cov";
+template <typename T_x, typename T_sigma, typename T_l, typename T_p,
+          require_all_stan_scalar_t<T_sigma, T_l, T_p>* = nullptr,
+          require_any_st_var<T_x, T_sigma, T_l, T_p>* = nullptr>
+Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic> gp_periodic_cov(
+    const std::vector<T_x>& x, T_sigma sigma, T_l l, T_p p) {
+  const char* fun = "gp_periodic_cov";
+
   check_positive(fun, "signal standard deviation", sigma);
   check_positive(fun, "length-scale", l);
   check_positive(fun, "period", p);
+
   size_t x_size = x.size();
   for (size_t i = 0; i < x_size; ++i) {
     check_not_nan(fun, "element of x", x[i]);
   }
 
-  Eigen::Matrix<var, -1, -1> cov(x_size, x_size);
   if (x_size == 0) {
-    return cov;
+    return Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic>();
   }
 
-  gp_periodic_cov_vari<T_x, var, var, var> *baseVari
-      = new gp_periodic_cov_vari<T_x, var, var, var>(x, sigma, l, p);
-
-  size_t pos = 0;
-  for (size_t j = 0; j < x_size; ++j) {
-    for (size_t i = (j + 1); i < x_size; ++i) {
-      cov.coeffRef(i, j).vi_ = baseVari->cov_lower_[pos];
-      cov.coeffRef(j, i).vi_ = cov.coeffRef(i, j).vi_;
-      ++pos;
+  double l_d = value_of(l);
+  double p_d = value_of(p);
+  double sigma_d = value_of(sigma);
+  double sigma_squared = sigma_d * sigma_d;
+  double pi_over_p = pi() / p_d;
+  double negative_two_over_l_squared = -2.0 / (l_d * l_d);
+
+  arena_matrix<Eigen::MatrixXd> arena_dist(x.size(), x.size());
+  arena_matrix<Eigen::MatrixXd> arena_sin_squared(x.size(), x.size());
+  arena_matrix<Eigen::MatrixXd> arena_sin_squared_derivative(x.size(),
+                                                             x.size());
+  arena_matrix<Eigen::MatrixXd> res_val(x.size(), x.size());
+
+  auto arena_x = to_arena_if<!is_constant<T_x>::value>(x);
```

Comment:
Another bug here.

Reply:
Should we put these bug fixes in another PR?

Reply:
Yeah, there's an issue for it here: https://github.com/stan-dev/math/issues/2025

I just made a comment here to remind me to fix this in the future (presumably whatever happens for #2025 will be suitable for this, or at least reveal a solution).